hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataInputNavigableJsonAdapter.java

```diff
@@ -88,34 +91,102 @@ private boolean isQuote() throws IOException {
+        private boolean hasLeftoverChar;
+        private int leftoverChar;
+
+        UTF8Reader(BufferObjectDataInput input) {
+            byte[] data = obtainBytes(input);
+            inputBuffer = ByteBuffer.wrap(data);
+            inputBuffer.position(input.position());
+            decoder = Thread.currentThread() instanceof OperationThread
+                    ? DECODER_THREAD_LOCAL.get()
+                    : Bits.UTF_8.newDecoder();
         }
 
+        // default read() implementation does not handle surrogate pairs well
+        @Override
+        public int read() throws IOException {
+            if (hasLeftoverChar) {
+                hasLeftoverChar = false;
+                return leftoverChar;
+            }
+            int charsRead = read(buffer, 0, 2);
+            switch (charsRead) {
+                case -1:
+                    return -1;
+                case 2:
+                    leftoverChar = buffer[1];
+                    hasLeftoverChar = true;
+                    return buffer[0];
+                case 1:
+                    return buffer[0];
+                default:
+                    throw new IllegalStateException("Unexpected result from read: " + charsRead);
+            }
+        }
+
+        @SuppressWarnings({"checkstyle:npathcomplexity", "checkstyle:cyclomaticcomplexity"})
         @Override
         public int read(char[] cbuf, int off, int len) throws IOException {
             if (off < 0 || (off + len) > cbuf.length) {
                 throw new IndexOutOfBoundsException();
             }
-            int i = 0;
-            try {
-                for (i = 0; i < len; i++) {
-                    byte firstByte = input.readByte();
-                    char c = Bits.readUtf8Char(input, firstByte);
-                    cbuf[off + i] = c;
+            if (len == 0) {
+                return 0;
+            }
+
+            int countRead = 0;
+            if (hasLeftoverChar) {
+                cbuf[off] = (char) leftoverChar;
+                hasLeftoverChar = false;
+                off++;
+                len--;
+                countRead++;
+            }
+            if (len == 0) {
+                return countRead;
+            }
+            if (!inputBuffer.hasRemaining()) {
+                return -1;
+            }
+            decoder.reset();
+            if (len == 1) {
+                int charRead = read();
+                if (charRead == -1) {
+                    return countRead == 0 ? -1 : countRead;
                 }
-            } catch (EOFException e) {
-                if (i == 0) {
-                    return -1;
+                cbuf[off] = (char) charRead;
+                return countRead + 1;
+            }
+
+            CharBuffer charbuffer = CharBuffer.wrap(cbuf, off, len);
+            CoderResult result = decoder.decode(inputBuffer, charbuffer, true);
+            if (result.isError()) {
+                result.throwException();
```

Comment:
What happens to the decoder state in this case? E.g. we try to decode the malformed input, receive an exception, then try to decode normal input from the same thread. Will it work well? And does it makes sense to add `Decoder.reset()`  to the `close` method?

Reply:
> we try to decode the malformed input, receive an exception, then try to decode normal input from the same thread

I don't think there's such a possibility. The `UTF8Reader` is instantiated with a specific byte buffer and `read` methods can only decode sequentially from that byte buffer; if position of byte buffer reaches a point of malformed input, every subsequent `read` will fail with the same exception.

> And does it makes sense to add Decoder.reset() to the close method?

According to the [decoding operation documentation](https://docs.oracle.com/javase/8/docs/api/index.html?java/nio/charset/CharsetDecoder.html) the decoder must be reset at start of decoding, as we do in implementation of `read(char[], int, int)`, so I don't think we need to reset the decoder in `close()`.